core-v1/src/main/scala/algebra/NonEmptyExprHList.scala

```diff
@@ -0,0 +1,113 @@
+package com.rallyhealth.vapors.v1
+
+package algebra
+
+import cats.arrow.Arrow
+import cats.implicits._
+import cats.{Functor, Semigroupal}
+import shapeless.ops.hlist.Comapped
+import shapeless.{::, HList, HNil}
+
+// TODO: Remove extra head param?
+sealed abstract class NonEmptyExprHList[-I, F[+_], +WL <: HList, +UL <: HList, OP[_]](
+  implicit
+  comapped: Comapped.Aux[WL, F, UL],
+) {
+
+  def ::[PI <: I, PH](expr: Expr[PI, F[PH], OP]): ExprHCons[PI, F, PH, WL, UL, OP] =
+    ExprHCons(expr, this)
+
+  def concatToHListWith[G[-_, +_]](v: ConcatToHList.Visitor[F, G, OP]): G[I, WL]
+
+  def zipToHListWith[G[-_, +_]](v: ZipToHList.Visitor[F, G, OP]): G[I, F[UL]]
```

Comment:
These should probably be prefixed by `visit` to match the `Expr.visit` method and avoid confusion when choosing a method to call from the DSL / builder side